Add bindings for calc()

Manishearth opened the pull request:
This adds bindings for dealing with refcounted calc() values.

So far we can get/set them, but we don't yet have the ability to copy them around. I'll work on that next.

Corresponding gecko bindings at https://github.com/servo/gecko-dev/compare/stylo...Manishearth:stylo-calc?expand=1

f? @emilio @heycam

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/12465)

<!-- Reviewable:end -->

Manishearth committed:
Regen bindings for Calc

Manishearth committed:
Add glue for calc values

Manishearth committed:
Allow calc() in GeckoStyleCoordConvertible

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/lib.rs, components/style/gecko_conversions.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!
* These commits modify style code, but no tests are modified. Please consider adding a test!

Manishearth commented on ports/geckolib/values.rs:
```diff
@@ -159,14 +159,16 @@ impl GeckoStyleCoordConvertible for LengthOrPercentage {
     fn to_gecko_style_coord(&self, unit: &mut nsStyleUnit, union: &mut nsStyleUnion) {
         match *self {
             LengthOrPercentage::Length(au) => {
+                unsafe { union.reset(unit) };
```
Perhaps I should move these reset calls out and remove the reset from set_calc_value. I designed it this way initially because set_calc_value was calling `nsStyleCoord::SetCalcValue()` which resets, but now it's a manual impl so this restriction doesn't apply anymore.

(I switched to a manual impl because in stylo we deal with `union,unit` pairs more than directly dealing with nsStyleCoord)

heycam replied:
Either way sounds OK to me, though if you pull the `reset()` call out please assert in `set_calc_value` that the unit is `eStyleUnit_Null`.

Alternatively, you could have a `Gecko_NewStyleCoordCalc` function that just creates the new refcounted `StyleCoordCalc` value and returns it, and `to_gecko_style_coord` here can set the unit and assign the `StyleCoordCalc` pointer into `union.mPointer`.  That would make all the branches of the match statement here line up.

Regardless, I think we'll need to make `StyleCoordValue` support thread-safe refcounting.  And we probably want to change all setter methods in `impl StyleCoordHelpers for nsStyleCoord` to do the right thing if the current (to be overwritten) value is a calc, and for `copy_from` to work if `other` is a calc too.

emilio commented:
The approach looks good to me, we should be extremely careful in order to avoid leaking values, but apart from that... :P

---

Reviewed 2 of 2 files at r1, 3 of 4 files at r2, 1 of 1 files at r3.
Review status: 6 of 7 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[components/style/gecko_conversions.rs, line 1 \[r3\]](https://reviewable.io:443/reviews/servo/servo/12465#-KMjQgDh0ss-ZfBxCVcj:-KMjQgDh0ss-ZfBxCVck:813801512) ([raw file](https://github.com/servo/servo/blob/59c7bc6fb1729190541132a54b8caba15a7b8f48/components/style/gecko_conversions.rs#L1)):*
> ```Rust
> //! This module contains conversion helpers between Servo and Gecko types
> ```

nit: License :)

---

*[ports/geckolib/values.rs, line 162 \[r3\]](https://reviewable.io:443/reviews/servo/servo/12465#-KMjAItgCjTnyP6VVqbh-r3-162:-KMjS2XjadEZ4m3oCInZ:-1095353534) ([raw file](https://github.com/servo/servo/blob/59c7bc6fb1729190541132a54b8caba15a7b8f48/ports/geckolib/values.rs#L162)):*
<details><summary><i>Previously, Manishearth (Manish Goregaokar) wrote…</i></summary>
> Perhaps I should move these reset calls out and remove the reset from set_calc_value. I designed it this way initially because set_calc_value was calling `nsStyleCoord::SetCalcValue()` which resets, but now it's a manual impl so this restriction doesn't apply anymore.
> 
> 
> 
> (I switched to a manual impl because in stylo we deal with `union,unit` pairs more than directly dealing with nsStyleCoord)
</details>
Yes, this sounds like it would simplify the logic a bit :)

---

*[ports/geckolib/values.rs, line 219 \[r3\]](https://reviewable.io:443/reviews/servo/servo/12465#-KMjRnHiMrQ5l4HYBkPh:-KMjRnHiMrQ5l4HYBkPi:-1105773524) ([raw file](https://github.com/servo/servo/blob/59c7bc6fb1729190541132a54b8caba15a7b8f48/ports/geckolib/values.rs#L219)):*
> ```Rust
>                 => Some(LengthOrPercentageOrAuto::Percentage(unsafe { *union.mFloat.as_ref() })),
>             nsStyleUnit::eStyleUnit_Calc
>                 => Some(LengthOrPercentageOrAuto::Calc(unsafe {union.get_calc().into()})),
> ```

nit: space

---

*[ports/geckolib/values.rs, line 243 \[r3\]](https://reviewable.io:443/reviews/servo/servo/12465#-KMjRac6SiJ5JPDSWqHl:-KMjRac6SiJ5JPDSWqHm:1516733647) ([raw file](https://github.com/servo/servo/blob/59c7bc6fb1729190541132a54b8caba15a7b8f48/ports/geckolib/values.rs#L243)):*
> ```Rust
>                 unsafe { *union.mInt.as_mut() = 0; }
>             },
>             LengthOrPercentageOrNone::Calc(calc) => unsafe {union.set_calc_value(unit, calc.into()) },
> ```

yeah, definitely  `set_calc_value` not implying `reset` could simplify a bit more the logic here :)

---

*[ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs, line 17 \[r3\]](https://reviewable.io:443/reviews/servo/servo/12465#-KMjQy64pqGUjUUKne4E:-KMjQy64pqGUjUUKne4F:-429357601) ([raw file](https://github.com/servo/servo/blob/59c7bc6fb1729190541132a54b8caba15a7b8f48/ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs#L17)):*
> ```Rust
>     /// Currently, this only happens if the nsStyleUnit
>     /// is a Calc
>     pub unsafe fn reset(&mut self, unit: &mut nsStyleUnit) {
> ```

We should try to not involve a FFI call always here, maybe could we worth to wrap in an `if`?

---

*[ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs, line 22 \[r3\]](https://reviewable.io:443/reviews/servo/servo/12465#-KMjSNfDHPUz-uJXAIhH:-KMjSNfDHPUz-uJXAIhI:1374343138) ([raw file](https://github.com/servo/servo/blob/59c7bc6fb1729190541132a54b8caba15a7b8f48/ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs#L22)):*
> ```Rust
> 
>     /// Set internal value to a calc() value
>     pub unsafe fn set_calc_value(&mut self, unit: &mut nsStyleUnit, v: nsStyleCoord_CalcValue) {
> ```

and if we're doing the `reset()` thing, maybe assert here that `unit != calc`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12465#-:-KMjSWmZHyKfchh7r9h7:852639617)*
<!-- Sent from Reviewable.io -->

heycam commented on ports/geckolib/gecko_bindings/tools/regen.py:
```diff
@@ -120,6 +120,7 @@
             "nsStyleEffects", "nsStyleImage", "nsStyleGradient",
             "nsStyleCoord", "nsStyleGradientStop", "nsStyleImageLayers",
             "nsStyleImageLayers::Layer", "nsStyleImageLayers::LayerType",
+             "nsStyleUnit", "nsStyleUnion", "nsStyleCoord::CalcValue",
```
Nit: this line is misaligned with the previous one.

heycam commented on components/style/gecko_conversions.rs:
```diff
@@ -0,0 +1,32 @@
+//! This module contains conversion helpers between Servo and Gecko types
+//! Ideally, it would be in geckolib itself, but coherence
+//! forces us to keep the traits and implementations here
```
Question: all uses of these two `From` impls are going to be in the geckolib crate, so why can't they live in ports/geckolib/values.rs?

emilio replied:
I *think* it is because you can only implement them either where `CalcLengthOrPercentage` is defined or where `nsStyleCoord_CalcValue` is defined, and doing it in `gecko_bindings` is impossible because is in the other way of the dependency chain.

Other solution would be to implement a custom trait as `GeckoStyleCoordConvertible` is. I'm fine with any of both options, either using the standard `From` and paying this price, or using a custom type :)

bholley replied:
I think I would prefer a custom trait if it means that we can keep this logic in geckolib.

heycam commented on ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs:
```diff
@@ -0,0 +1,29 @@
+use structs::{nsStyleCoord_CalcValue, nsStyleCoord_Calc, nsStyleUnit, nsStyleUnion};
+use bindings::{Gecko_ResetStyleCoord, Gecko_SetStyleCoordCalcValue};
+
+// Functions here are unsafe because it is possible to use the wrong nsStyleUnit
```
I'd like to understand the difference between `unsafe` on the function interface vs wrapping the body of the function in `unsafe` here.  Is it because we have a requirement that is not expressible in the type system, which must be adhered to otherwise we will have a memory safety issue, namely that you must pass in a reference to the `nsStyleUnit` associated with the `nsStyleUnion` and not just any one you have access to?  And to signal to users of `nsStyleUnion` that we have other requirements the compiler won't catch, we use `unsafe` here?

Manishearth replied:
Basically yes. I will refactor this in a later PR to fix it.

Manishearth commented:
Should be done now.

r? @emilio or @heycam

Manishearth committed:
Handle Calc refcounting

Manishearth commented:
I tried adding a Drop impl to nsStyleCoord, but it turns out we let Gecko handle drops so that just leads to Servo double-dropping it and causing havoc.

Manishearth commented:
Gecko bug at https://bugzilla.mozilla.org/show_bug.cgi?id=1287435

emilio commented:
The rest of the logic looks sound to me, though @heycam might want to take another look.

-S-awaiting-review +S-needs-code-changes 

---

Reviewed 1 of 4 files at r2, 1 of 5 files at r5, 3 of 4 files at r6.
Review status: 5 of 8 files reviewed at latest revision, 10 unresolved discussions.

---

*[ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs, line 28 \[r7\]](https://reviewable.io:443/reviews/servo/servo/12465#-KMyt1u84FH9Ror-EeAg:-KMyt1u9b1NJMZb9OblK:-606173492) ([raw file](https://github.com/servo/servo/blob/7f63fca50675f9aafcf0f6fd0e454a7dbc3546da/ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs#L28)):*
> ```Rust
>     pub unsafe fn set_calc_value(&mut self, unit: &mut nsStyleUnit, v: nsStyleCoord_CalcValue) {
>         // Calc should have been cleaned up
>         assert!(*unit != nsStyleUnit::eStyleUnit_Calc);
> ```

nit: we tend to only use `debug_assert!` in geckolib, sorry for not having being more explicit before.

---

*[ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs, line 32 \[r7\]](https://reviewable.io:443/reviews/servo/servo/12465#-KMytUXeWaeGPBzW2mH_:-KMytUXeWaeGPBzW2mHa:-133989733) ([raw file](https://github.com/servo/servo/blob/7f63fca50675f9aafcf0f6fd0e454a7dbc3546da/ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs#L32)):*
> ```Rust
>     }
> 
>     pub unsafe fn get_calc(&self) -> nsStyleCoord_CalcValue {
> ```

I'd really love having more assertions on this methods, but passing the unit around might be not convenient?

I really would not matter having to pass the unit here, but if it's too much of a hassle I guess not doing it is fine.

---

*[ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs, line 36 \[r7\]](https://reviewable.io:443/reviews/servo/servo/12465#-KMys5aUAOg-ay573cJw:-KMys5aUAOg-ay573cJx:-1681201013) ([raw file](https://github.com/servo/servo/blob/7f63fca50675f9aafcf0f6fd0e454a7dbc3546da/ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs#L36)):*
> ```Rust
>     }
> 
>     pub unsafe fn addref_opt(&mut self, unit: &nsStyleUnit) {
> ```

nit: Can we call this `addref_if_calc`? It's not evident otherwise when reading the callers.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12465#-:-KMyuFcUSwUU9dnwlSEE:-23778177)*
<!-- Sent from Reviewable.io -->

emilio commented:
Reviewed 2 of 5 files at r5, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[components/style/gecko_conversions.rs, line 17 \[r7\]](https://reviewable.io:443/reviews/servo/servo/12465#-KMyueVzeCfKgrddVjI4:-KMyueW2cR3hp-joXRpY:-1104124876) ([raw file](https://github.com/servo/servo/blob/7f63fca50675f9aafcf0f6fd0e454a7dbc3546da/components/style/gecko_conversions.rs#L17)):*
> ```Rust
>         let has_percentage = other.percentage.is_some();
>         nsStyleCoord_CalcValue {
>             mLength: other.length.map(|l| l.0).unwrap_or(0),
> ```

nit: `map_or`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12465)*
<!-- Sent from Reviewable.io -->

bholley commented on ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs:
```diff
@@ -0,0 +1,54 @@
+// Functions here are unsafe because it is possible to use the wrong nsStyleUnit
+// FIXME we should be pairing up nsStyleUnion and nsStyleUnit somehow
+// nsStyleCoord is one way to do it, but there are other structs using pairs
+// of union and unit too
```
Yeah, the main reason for the separation is that in Gecko, we sometimes have tuples of these things, and the unions and units are stored in separate arrays. Maybe we could have some sort of temporary struct that we synthesize (unsafely) and pass around (safely)? Would that impose a performance cost?

Manishearth replied:
Yes, this is the plan. Shouldn't have a perf cost (because we pass them in as arguments anyway); and I bet llvm will optimize it away even if it did.

A slight issue with this abstraction is that it's still technically unsafe because folks can always modify the values in nsStyleCoord directly. I could try to make those private via bindgen stuff, though.

bholley replied:
I was imagining that we'd create a some temporary tuple struct that isn't actually nsStyleCoord, but if we can make that nsStyleCoord itself with similar properties, that's even better.

bholley commented on ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs:
```diff
@@ -0,0 +1,54 @@
+    /// reset() the union before calling this
+    pub unsafe fn set_calc_value(&mut self, unit: &mut nsStyleUnit, v: nsStyleCoord_CalcValue) {
+        // Calc should have been cleaned up
+        assert!(*unit != nsStyleUnit::eStyleUnit_Calc);
```
We should use debug_assert everywhere.

bholley commented on ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs:
```diff
@@ -0,0 +1,54 @@
+        (*self.as_calc())._base
+    }
+
+    pub unsafe fn addref_opt(&mut self, unit: &nsStyleUnit) {
```
What does the _opt mean here? Optional? Seems kinda ambiguous.

But also, maybe this is the wrong level of abstraction? Shouldn't we just have a set() for nsStyleUnion that does the refcount munging as appropriate, and possibly a set_no_calc that asserts against calc in the cases we know it can't happen?

bholley replied:
Though I guess the issue here is that this needs the tuple abstraction layer we talked about above.

Manishearth replied:
Yeah, this is all temporary. I will be doing that tuple layer in a different PR, and it will have set_ methods and stuff. I'm trying to keep my PRs which span repos to be minimal in size.

heycam commented on ports/geckolib/values.rs:
```diff
@@ -56,6 +56,7 @@ impl StyleCoordHelpers for nsStyleCoord {
 
     #[inline]
     fn set_auto(&mut self) {
+        unsafe { self.mValue.reset(&mut self.mUnit)};
```
Nit: space before `}`, and in all the other methods in this patch.

heycam commented on ports/geckolib/values.rs:
```diff
@@ -166,7 +174,7 @@ impl GeckoStyleCoordConvertible for LengthOrPercentage {
                 unsafe { *union.mFloat.as_mut() = p; }
             },
-            LengthOrPercentage::Calc(_) => unimplemented!(),
+            LengthOrPercentage::Calc(calc) => unsafe {union.set_calc_value(unit, calc.into()) },
```
Nit: space after `{`.

heycam commented on ports/geckolib/values.rs:
```diff
@@ -177,14 +185,15 @@ impl GeckoStyleCoordConvertible for LengthOrPercentage {
                 => Some(LengthOrPercentage::Percentage(unsafe { *union.mFloat.as_ref() })),
             nsStyleUnit::eStyleUnit_Calc
-                => unimplemented!(),
+                => Some(LengthOrPercentage::Calc(unsafe { union.get_calc().into()})),
```
Nit: space before `}`.

heycam commented on ports/geckolib/values.rs:
```diff
@@ -157,6 +164,7 @@ pub trait GeckoStyleCoordConvertible : Sized {
 
 impl GeckoStyleCoordConvertible for LengthOrPercentage {
     fn to_gecko_style_coord(&self, unit: &mut nsStyleUnit, union: &mut nsStyleUnion) {
```
If the functions on `nsStyleUnion` are unsafe because they require particular `nsStyleUnit` references to be passed in, shouldn't that also mean `to_gecko_style_coord` and `from_gecko_style_coord` are unsafe too?  Or do I misunderstand the use of unsafe on the `nsStyleUnion` functions.

heycam commented on ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs:
```diff
@@ -29,6 +30,25 @@ impl nsStyleUnion {
+        }
+    }
+
+    pub unsafe fn as_calc_mut(&mut self) -> &mut nsStyleCoord_Calc {
```
Do `as_calc_mut` and `as_calc` need to be `pub`?  Looks like they're not called from elsewhere.

heycam commented on ports/geckolib/properties.mako.rs:
```diff
@@ -317,9 +317,10 @@ def set_gecko_property(ffi_name, expr):
+        unsafe { self.gecko.${union_ffi_name}.reset(&mut self.gecko.${unit_ffi_name}) };
         self.gecko.${unit_ffi_name} =  other.gecko.${unit_ffi_name};
         self.gecko.${union_ffi_name} = other.gecko.${union_ffi_name};
+        unsafe { self.gecko.${union_ffi_name}.addref_opt(&self.gecko.${unit_ffi_name}) };
```
Rather than assigning directly to the unit and union, and then calling `addref_opt` (or `addref_if_calc` if renamed), we could just have a single call to a `set` function on `nsStyleUnion` that takes a reference to the unit as an argument, and does the unit check and addref if necessary.  Currently it's kind of splitting the responsibility of dealing with different units correctly across `nsStyleUnion` functions and code here poking directly into the unit/union.

heycam commented on ports/geckolib/properties.mako.rs:
```diff
@@ -347,12 +348,14 @@ ${impl_split_style_coord(ident,
         self.gecko.${y_unit_ffi_name} = other.gecko.${y_unit_ffi_name};
         self.gecko.${y_union_ffi_name} = other.gecko.${y_union_ffi_name};
+        unsafe { self.gecko.${x_union_ffi_name}.addref_opt(&self.gecko.${x_unit_ffi_name}) };
+        unsafe { self.gecko.${y_union_ffi_name}.addref_opt(&self.gecko.${y_unit_ffi_name}) };
```
Similarly here I think it would encapsulate the behaviour better to have a `copy_from` function on `nsStyleUnion` that deals with addrefing if needed.

heycam commented on ports/geckolib/properties.mako.rs:
```diff
@@ -628,9 +631,10 @@ fn static_assert() {
+        unsafe { self.gecko.mZIndex.mValue.reset(&mut self.gecko.mZIndex.mUnit) };
         self.gecko.mZIndex.mUnit = other.gecko.mZIndex.mUnit;
         self.gecko.mZIndex.mValue = other.gecko.mZIndex.mValue;
+        unsafe { self.gecko.mZIndex.mValue.addref_opt(&self.gecko.mZIndex.mUnit) };
```
z-index doesn't take calc() values, so instead you could assert that the new unit value isn't calc.

Manishearth committed:
Address review comments

Manishearth commented:
Addressed most of the issues

---

Review status: 4 of 8 files reviewed at latest revision, 23 unresolved discussions.

---

*[components/style/gecko_conversions.rs, line 3 \[r2\]](https://reviewable.io:443/reviews/servo/servo/12465#-KMjAItgCjTnyP6VVqbf-r2-3:-KN0a1yRD5_B0KkAnZ0L:2010699936) ([raw file](https://github.com/servo/servo/blob/b130f3af2cab019e4423ce3cda9f53175a88e3d9/components/style/gecko_conversions.rs#L3)):*
<details><summary><i>Previously, bholley (Bobby Holley) wrote…</i></summary>
> I think I would prefer a custom trait if it means that we can keep this logic in geckolib.
</details>
The advantage of From is that it automatically gets you into and some other helpers, and the trait is autoimported.

---

*[ports/geckolib/properties.mako.rs, line 323 \[r7\]](https://reviewable.io:443/reviews/servo/servo/12465#-KMyCd_lUmznR4oR-gjJ-r7-323:-KN0Zcui0D7cpDZ0Hc9o:208602143) ([raw file](https://github.com/servo/servo/blob/38206b1117e9ec33c8b63f7728d3f5fff73a88aa/ports/geckolib/properties.mako.rs#L323)):*
<details><summary><i>Previously, heycam (Cameron McCormack) wrote…</i></summary>
> Rather than assigning directly to the unit and union, and then calling `addref_opt` (or `addref_if_calc` if renamed), we could just have a single call to a `set` function on `nsStyleUnion` that takes a reference to the unit as an argument, and does the unit check and addref if necessary.  Currently it's kind of splitting the responsibility of dealing with different units correctly across `nsStyleUnion` functions and code here poking directly into the unit/union.
</details>
This will happen in my next PR when I abstractify all the things

---

*[ports/geckolib/values.rs, line 166 \[r4\]](https://reviewable.io:443/reviews/servo/servo/12465#-KMjAItgCjTnyP6VVqbh-r4-166:-KN0ZJDcjTPlUjEFJ5G3:1660649091) ([raw file](https://github.com/servo/servo/blob/7f4628535aa9bdf70579f5b072218c1a42a4a27c/ports/geckolib/values.rs#L166)):*
<details><summary><i>Previously, heycam (Cameron McCormack) wrote…</i></summary>
> If the functions on `nsStyleUnion` are unsafe because they require particular `nsStyleUnit` references to be passed in, shouldn't that also mean `to_gecko_style_coord` and `from_gecko_style_coord` are unsafe too?  Or do I misunderstand the use of unsafe on the `nsStyleUnion` functions.
</details>
They are, but this is temporary, so I didn't want to infect everything with unsafe

---

*[ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs, line 32 \[r7\]](https://reviewable.io:443/reviews/servo/servo/12465#-KMytUXeWaeGPBzW2mH_:-KN0Z95N31DrVTMepdD5:208602143) ([raw file](https://github.com/servo/servo/blob/38206b1117e9ec33c8b63f7728d3f5fff73a88aa/ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs#L32)):*
<details><summary><i>Previously, emilio (Emilio Cobos Álvarez) wrote…</i></summary>
> I'd really love having more assertions on this methods, but passing the unit around might be not convenient?
> 
> I really would not matter having to pass the unit here, but if it's too much of a hassle I guess not doing it is fine.
</details>
This will happen in my next PR when I abstractify all the things

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12465#-:-KN0_yVS7IG11m-z-nOu:508533415)*
<!-- Sent from Reviewable.io -->

Manishearth committed:
Sync stylo bindings

Manishearth commented:
Updated

Manishearth commented:
r? @heycam

heycam commented:
OK, looks good to me (though good to get the tuple abstraction done soon), but check @bholley is happy with the `From` impls before landing.

@bors-servo delegate+

bors-servo commented:
:v: @Manishearth can now approve this pull request

bholley commented:
Given that @SimonSapin prefers to shift towards more conditional compilation and less traits in the style system, I'm fine with using From here.

Manishearth commented:
@bors-servo r=heycam 

I'll get to work on the tuple abstraction (or whatever) tomorrow then.

For Maximum Safety we might need to tweak bindgen so that nsstyleunion/unit can be made private fields.

bors-servo commented:
:pushpin: Commit f51b115 has been approved by `heycam`

<!-- @bors-servo r=heycam f51b115db0e3ab9973b1e78927ca23fb81f45ed8 -->

bors-servo commented:
:hourglass: Testing commit f51b115 with merge 2e03e55...

larsbergstrom commented:
infra issues; will reopen and retry once they are fixed.

larsbergstrom commented:
@bors-servo retry clean

larsbergstrom commented:
@bors-servo retry force clean p=1

bors-servo commented:
:hourglass: Testing commit f51b115 with merge 3760440...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/1960), [arm32](http://build.servo.org/builders/arm32/builds/2071), [arm64](http://build.servo.org/builders/arm64/builds/2072), [linux-dev](http://build.servo.org/builders/linux-dev/builds/2115), [linux-rel](http://build.servo.org/builders/linux-rel/builds/2235), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/2144), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/2131), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/2203), [windows-dev](http://build.servo.org/builders/windows-dev/builds/17)
